base/utf16.jl

```diff
@@ -71,25 +152,346 @@ function encode16(s::AbstractString)
+                    if ch > 0x7ff
+                        cnt3 += 1
+                    elseif ch > 0x7f
+                        cnt2 += 1
+                    end
+                else
+                    utf_errfunc(UTF_ERR_LONG, pos-2, ch)
+                end
+            else
+                utf_errfunc(UTF_ERR_INVALID, pos, ch)
+            end
+        end
+    end
+    cntT, cnt2, cnt3, cnt4, flags | (cnt3 == 0 ? 0 : UTF_UNICODE3) | (cnt4 == 0 ? 0 : UTF_UNICODE4)
+end
+
+#=
+@doc """
+@brief      Validates and calculates number of characters in UTF-16 string
+
+@param[in]  iStr::UTF16String
+@param[in]  iOptions flags to determine error handling (default 0)
+
+@return     (total characters, 2-byte, 3-byte, 4-byte, flags)
+@throws     UnicodeError
+""" ->
+=#
+function utf16_validate(iStr::Array{UInt16,1}, iOptions::Integer=0)
+    ch::UInt32 = 0
+    cntT::UInt = 0      # total # of characters
+    cnt2::UInt = 0      # number of characters in the range (0x80-0x7ff)
+    cnt3::UInt = 0      # number of characters in the range (0x800-0xd7ff,0xe000-0xffff)
+    cnt4::UInt = 0      # number of characters in the range (0x10000-0x10ffff)
+    flags::UInt = 0
+    pos::UInt = 0
+    len::UInt = sizeof(iStr) >>> 1
+    @inbounds iStr[len] == 0 && (len -= 1)
+    while pos < len
+        @inbounds ch = iStr[pos += 1]
+        cntT += 1
+        if ch > 0x7f
+            if ch < 0x100
+                cnt2 += 1
+                flags |= UTF_LATIN1
+            elseif ch < 0x800
+                cnt2 += 1
+                flags |= UTF_UNICODE2
+            elseif !is_surrogate_char(ch)
+                cnt3 += 1
+            elseif is_surrogate_lead(ch)
+                pos == len && utf_errfunc(UTF_ERR_MISSING_SURROGATE, pos, ch)
+                # next character *must* be a trailing surrogate character
+                @inbounds ch = iStr[pos += 1]
+                !is_surrogate_trail(ch) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos, ch)
+                cnt4 += 1
+            else
+                utf_errfunc(UTF_ERR_NOT_LEAD, pos, ch)
+            end
+        end
+    end
+    cntT, cnt2, cnt3, cnt4, flags | (cnt3 == 0 ? 0 : UTF_UNICODE3) | (cnt4 == 0 ? 0 : UTF_UNICODE4)
+end
+
+#=
+"""
+@brief      Validates and calculates number of characters in UTF-32 string
+
+@param[in]  iStr::UTF32String
+@param[in]  iOptions flags to determine error handling (default 0)
+
+@return     (total characters, 2-byte, 3-byte, 4-byte, flags)
+@throws     UnicodeError
+"""
+=#
+function utf32_validate(iStr::Array{Char,1}, iOptions::Integer=0)
+    cntT::UInt = 0      # total # of characters
+    cnt2::UInt = 0      # number of characters in the range (0x80-0x7ff)
+    cnt3::UInt = 0      # number of characters in the range (0x800-0xd7ff,0xe000-0xffff)
+    cnt4::UInt = 0      # number of characters in the range (0x10000-0x10ffff)
+    pos::UInt = 0
+    flags::UInt = 0
+    len::UInt = (sizeof(iStr) >>> 2)
+    @inbounds iStr[len] == 0 && (len -= 1)
```

Comment:
null-termination isn't optional on a utf32 (or utf16) string. making it conditional would imply this could accidentally drop a terminal null byte that was actually part of the string.